// drizzle-orm-mysql/mysql2/driver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.drizzle = exports.MySqlDatabase = exports.MySql2Driver = void 0;
const dialect_1 = require("../dialect");
const session_1 = require("./session");
class MySql2Driver {
    constructor(client, dialect, options = {}) {
        this.client = client;
        this.dialect = dialect;
        this.options = options;
    }
    createSession() {
        return new session_1.MySql2Session(this.client, this.dialect, { logger: this.options.logger });
    }
}
exports.MySql2Driver = MySql2Driver;
var db_1 = require("../db");
Object.defineProperty(exports, "MySqlDatabase", { enumerable: true, get: function () { return db_1.MySqlDatabase; } });
function drizzle(client, config = {}) {
    const dialect = new dialect_1.MySqlDialect();
    const driver = new MySql2Driver(client, dialect, { logger: config.logger });
    const session = driver.createSession();
    return dialect.createDB(session);
}
exports.drizzle = drizzle;
