// drizzle-orm-pg/db.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PgDatabase = void 0;
const query_builders_1 = require("./query-builders");
const table_1 = require("./table");
const utils_1 = require("./utils");
class PgDatabase {
    constructor(
    /** @internal */
    dialect, 
    /** @internal */
    session) {
        this.dialect = dialect;
        this.session = session;
    }
    select(from) {
        const fields = (0, utils_1.orderSelectedFields)(from[table_1.PgTable.Symbol.Columns]);
        return new query_builders_1.PgSelect(from, fields, this.session, this.dialect);
    }
    update(table) {
        return new query_builders_1.PgUpdateBuilder(table, this.session, this.dialect);
    }
    insert(table) {
        return new query_builders_1.PgInsertBuilder(table, this.session, this.dialect);
    }
    delete(table) {
        return new query_builders_1.PgDelete(table, this.session, this.dialect);
    }
    execute(query) {
        return this.session.execute(query.getSQL());
    }
}
exports.PgDatabase = PgDatabase;
