// drizzle-orm-pg/foreign-keys.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.foreignKey = exports.ForeignKey = exports.ForeignKeyBuilder = void 0;
const table_1 = require("./table");
class ForeignKeyBuilder {
    constructor(config, actions) {
        /** @internal */
        this._onUpdate = 'no action';
        /** @internal */
        this._onDelete = 'no action';
        this.reference = () => {
            const { columns, foreignColumns } = config();
            return { columns, foreignTable: foreignColumns[0].table, foreignColumns };
        };
        if (actions) {
            this._onUpdate = actions.onUpdate;
            this._onDelete = actions.onDelete;
        }
    }
    onUpdate(action) {
        this._onUpdate = typeof action === 'undefined' ? 'no action' : action;
        return this;
    }
    onDelete(action) {
        this._onDelete = typeof action === 'undefined' ? 'no action' : action;
        return this;
    }
    /** @internal */
    build(table) {
        return new ForeignKey(table, this);
    }
}
exports.ForeignKeyBuilder = ForeignKeyBuilder;
class ForeignKey {
    constructor(table, builder) {
        this.table = table;
        this.reference = builder.reference;
        this.onUpdate = builder._onUpdate;
        this.onDelete = builder._onDelete;
    }
    getName() {
        const { columns, foreignColumns } = this.reference();
        const columnNames = columns.map((column) => column.name);
        const foreignColumnNames = foreignColumns.map((column) => column.name);
        const chunks = [
            this.table[table_1.PgTable.Symbol.Name],
            ...columnNames,
            foreignColumns[0].table[table_1.PgTable.Symbol.Name],
            ...foreignColumnNames,
        ];
        return `${chunks.join('_')}_fk`;
    }
}
exports.ForeignKey = ForeignKey;
function foreignKey(config) {
    function mappedConfig() {
        const { columns, foreignColumns } = config;
        return {
            columns,
            foreignColumns,
        };
    }
    return new ForeignKeyBuilder(mappedConfig);
}
exports.foreignKey = foreignKey;
