// drizzle-orm-pg/neondb/driver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.drizzle = exports.NeonDriver = void 0;
const serverless_1 = require("@neondatabase/serverless");
const db_1 = require("../db");
const dialect_1 = require("../dialect");
const session_1 = require("./session");
class NeonDriver {
    constructor(client, dialect, options = {}) {
        this.client = client;
        this.dialect = dialect;
        this.options = options;
        this.initMappers();
    }
    createSession() {
        return new session_1.NeonSession(this.client, this.dialect, { logger: this.options.logger });
    }
    initMappers() {
        serverless_1.types.setTypeParser(serverless_1.types.builtins.TIMESTAMPTZ, (val) => val);
        serverless_1.types.setTypeParser(serverless_1.types.builtins.TIMESTAMP, (val) => val);
        serverless_1.types.setTypeParser(serverless_1.types.builtins.DATE, (val) => val);
    }
}
exports.NeonDriver = NeonDriver;
function drizzle(client, config = {}) {
    const dialect = new dialect_1.PgDialect();
    const driver = new NeonDriver(client, dialect, { logger: config.logger });
    const session = driver.createSession();
    return new db_1.PgDatabase(dialect, session);
}
exports.drizzle = drizzle;
