// drizzle-orm-pg/neondb/session.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NeonSession = exports.NeonPreparedQuery = void 0;
const drizzle_orm_1 = require("drizzle-orm");
const sql_1 = require("drizzle-orm/sql");
const utils_1 = require("drizzle-orm/utils");
const session_1 = require("../session");
class NeonPreparedQuery extends session_1.PreparedQuery {
    constructor(client, queryString, params, logger, fields, name) {
        super();
        this.client = client;
        this.params = params;
        this.logger = logger;
        this.fields = fields;
        this.rawQuery = {
            name,
            text: queryString,
        };
        this.query = {
            name,
            text: queryString,
            rowMode: 'array',
        };
    }
    execute(placeholderValues = {}) {
        const params = (0, sql_1.fillPlaceholders)(this.params, placeholderValues);
        this.logger.logQuery(this.rawQuery.text, params);
        const { fields } = this;
        if (!fields) {
            return this.client.query(this.rawQuery, params);
        }
        const result = this.client.query(this.query, params);
        return result.then((result) => result.rows.map((row) => (0, utils_1.mapResultRow)(fields, row)));
    }
    all(placeholderValues = {}) {
        const params = (0, sql_1.fillPlaceholders)(this.params, placeholderValues);
        this.logger.logQuery(this.rawQuery.text, params);
        return this.client.query(this.rawQuery, params).then((result) => result.rows);
    }
    values(placeholderValues = {}) {
        const params = (0, sql_1.fillPlaceholders)(this.params, placeholderValues);
        this.logger.logQuery(this.rawQuery.text, params);
        return this.client.query(this.query, params).then((result) => result.rows);
    }
}
exports.NeonPreparedQuery = NeonPreparedQuery;
class NeonSession extends session_1.PgSession {
    constructor(client, dialect, options = {}) {
        var _a;
        super(dialect);
        this.client = client;
        this.logger = (_a = options.logger) !== null && _a !== void 0 ? _a : new drizzle_orm_1.NoopLogger();
    }
    prepareQuery(query, fields, name) {
        return new NeonPreparedQuery(this.client, query.sql, query.params, this.logger, fields, name);
    }
    query(query, params) {
        return __awaiter(this, void 0, void 0, function* () {
            this.logger.logQuery(query, params);
            const result = yield this.client.query({
                rowMode: 'array',
                text: query,
                values: params,
            });
            return result;
        });
    }
    queryObjects(query, params) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.client.query(query, params);
        });
    }
}
exports.NeonSession = NeonSession;
