// drizzle-orm-pg/query-builders/insert.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PgInsert = exports.PgInsertBuilder = void 0;
const drizzle_orm_1 = require("drizzle-orm");
const query_promise_1 = require("drizzle-orm/query-promise");
const sql_1 = require("drizzle-orm/sql");
const table_1 = require("../table");
const utils_1 = require("../utils");
class PgInsertBuilder {
    constructor(table, session, dialect) {
        this.table = table;
        this.session = session;
        this.dialect = dialect;
    }
    values(...values) {
        const mappedValues = values.map((entry) => {
            const result = {};
            const cols = this.table[drizzle_orm_1.Table.Symbol.Columns];
            for (const colKey of Object.keys(entry)) {
                const colValue = entry[colKey];
                if (colValue instanceof sql_1.SQL) {
                    result[colKey] = colValue;
                }
                else {
                    result[colKey] = new sql_1.Param(colValue, cols[colKey]);
                }
            }
            return result;
        });
        return new PgInsert(this.table, mappedValues, this.session, this.dialect);
    }
}
exports.PgInsertBuilder = PgInsertBuilder;
class PgInsert extends query_promise_1.QueryPromise {
    constructor(table, values, session, dialect) {
        super();
        this.session = session;
        this.dialect = dialect;
        this.execute = (placeholderValues) => {
            return this._prepare().execute(placeholderValues);
        };
        this.config = { table, values };
    }
    returning(fields = this.config.table[table_1.PgTable.Symbol.Columns]) {
        this.config.returning = (0, utils_1.orderSelectedFields)(fields);
        return this;
    }
    onConflictDoNothing(config = {}) {
        if (config.target === undefined) {
            this.config.onConflict = (0, sql_1.sql) `do nothing`;
        }
        else {
            let targetColumn = '';
            if (Array.isArray(config.target)) {
                targetColumn = config.target.map((it) => this.dialect.escapeName(it.name)).join(',');
            }
            else {
                targetColumn = this.dialect.escapeName(config.target.name);
            }
            const whereSql = config.where ? (0, sql_1.sql) ` where ${config.where}` : (0, sql_1.sql) ``;
            this.config.onConflict = (0, sql_1.sql) `(${sql_1.sql.raw(targetColumn)})${whereSql} do nothing`;
        }
        return this;
    }
    onConflictDoUpdate(config) {
        const whereSql = config.where ? (0, sql_1.sql) ` where ${config.where}` : (0, sql_1.sql) ``;
        const setSql = this.dialect.buildUpdateSet(this.config.table, (0, utils_1.mapUpdateSet)(this.config.table, config.set));
        let targetColumn = '';
        if (Array.isArray(config.target)) {
            targetColumn = config.target.map((it) => this.dialect.escapeName(it.name)).join(',');
        }
        else {
            targetColumn = this.dialect.escapeName(config.target.name);
        }
        this.config.onConflict = (0, sql_1.sql) `(${sql_1.sql.raw(targetColumn)})${whereSql} do update set ${setSql}`;
        return this;
    }
    /** @internal */
    getSQL() {
        return this.dialect.buildInsertQuery(this.config);
    }
    toSQL() {
        return this.dialect.sqlToQuery(this.getSQL());
    }
    _prepare(name) {
        return this.session.prepareQuery(this.toSQL(), this.config.returning, name);
    }
    prepare(name) {
        return this._prepare(name);
    }
}
exports.PgInsert = PgInsert;
