// drizzle-orm-pg/query-builders/update.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PgUpdate = exports.PgUpdateBuilder = void 0;
const query_promise_1 = require("drizzle-orm/query-promise");
const table_1 = require("../table");
const utils_1 = require("../utils");
class PgUpdateBuilder {
    constructor(table, session, dialect) {
        this.table = table;
        this.session = session;
        this.dialect = dialect;
    }
    set(values) {
        return new PgUpdate(this.table, (0, utils_1.mapUpdateSet)(this.table, values), this.session, this.dialect);
    }
}
exports.PgUpdateBuilder = PgUpdateBuilder;
class PgUpdate extends query_promise_1.QueryPromise {
    constructor(table, set, session, dialect) {
        super();
        this.session = session;
        this.dialect = dialect;
        this.execute = (placeholderValues) => {
            return this._prepare().execute(placeholderValues);
        };
        this.config = { set, table };
    }
    where(where) {
        this.config.where = where;
        return this;
    }
    returning(fields = this.config.table[table_1.PgTable.Symbol.Columns]) {
        this.config.returning = (0, utils_1.orderSelectedFields)(fields);
        return this;
    }
    /** @internal */
    getSQL() {
        return this.dialect.buildUpdateQuery(this.config);
    }
    toSQL() {
        return this.dialect.sqlToQuery(this.getSQL());
    }
    _prepare(name) {
        return this.session.prepareQuery(this.toSQL(), this.config.returning, name);
    }
    prepare(name) {
        return this._prepare(name);
    }
}
exports.PgUpdate = PgUpdate;
