// drizzle-orm-mysql/query-builders/delete.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MySqlDelete = void 0;
const query_promise_1 = require("drizzle-orm/query-promise");
class MySqlDelete extends query_promise_1.QueryPromise {
    constructor(table, session, dialect) {
        super();
        this.table = table;
        this.session = session;
        this.dialect = dialect;
        this.execute = (placeholderValues) => {
            return this._prepare().execute(placeholderValues);
        };
        this.config = { table };
    }
    where(where) {
        this.config.where = where;
        return this;
    }
    // returning(): Omit<MySqlDelete<TTable, InferModel<TTable>>, 'where' | 'returning'>;
    // returning<TSelectedFields extends SelectFields>(
    // 	fields: TSelectedFields,
    // ): Omit<MySqlDelete<TTable, SelectResultFields<TSelectedFields>>, 'where' | 'returning'>;
    // returning(
    // 	fields: SelectFields = this.config.table[MySqlTable.Symbol.Columns],
    // ): Omit<MySqlDelete<TTable, any>, 'where' | 'returning'> {
    // 	this.config.returning = orderSelectedFields(fields);
    // 	return this;
    // }
    /** @internal */
    getSQL() {
        return this.dialect.buildDeleteQuery(this.config);
    }
    toSQL() {
        return this.dialect.sqlToQuery(this.getSQL());
    }
    _prepare(name) {
        return this.session.prepareQuery(this.toSQL(), this.config.returning, name);
    }
    prepare(name) {
        return this._prepare(name);
    }
}
exports.MySqlDelete = MySqlDelete;
