// drizzle-orm-pg/columns/bigint.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.bigint = exports.PgBigInt64 = exports.PgBigInt64Builder = exports.PgBigInt53 = exports.PgBigInt53Builder = void 0;
const common_1 = require("./common");
class PgBigInt53Builder extends common_1.PgColumnBuilder {
    /** @internal */
    build(table) {
        return new PgBigInt53(table, this.config);
    }
}
exports.PgBigInt53Builder = PgBigInt53Builder;
class PgBigInt53 extends common_1.PgColumn {
    getSQLType() {
        return 'bigint';
    }
    mapFromDriverValue(value) {
        if (typeof value === 'number') {
            return value;
        }
        return parseInt(value);
    }
}
exports.PgBigInt53 = PgBigInt53;
class PgBigInt64Builder extends common_1.PgColumnBuilder {
    /** @internal */
    build(table) {
        return new PgBigInt64(table, this.config);
    }
}
exports.PgBigInt64Builder = PgBigInt64Builder;
class PgBigInt64 extends common_1.PgColumn {
    getSQLType() {
        return 'bigint';
    }
    mapFromDriverValue(value) {
        return BigInt(value);
    }
}
exports.PgBigInt64 = PgBigInt64;
function bigint(name, config) {
    if (config.mode === 'number') {
        return new PgBigInt53Builder(name);
    }
    return new PgBigInt64Builder(name);
}
exports.bigint = bigint;
