// drizzle-orm-pg/columns/custom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.customType = void 0;
const common_1 = require("./common");
function returnColumn(table, config, sqlName, mapTo, mapFrom) {
    return new class extends common_1.PgColumn {
        getSQLType() {
            return sqlName;
        }
        mapFromDriverValue(value) {
            if (typeof mapFrom === 'function') {
                return mapFrom(value);
            }
            else {
                return value;
            }
        }
        mapToDriverValue(value) {
            if (typeof mapTo === 'function') {
                return mapTo(value);
            }
            else {
                return value;
            }
        }
    }(table, config);
}
/**
 * Custom postgresql database data type generator
 */
function customType(customTypeParams) {
    return (dbName, fieldConfig) => new class extends common_1.PgColumnBuilder {
        /** @internal */
        build(table) {
            return returnColumn(table, this.config, customTypeParams.dataType(fieldConfig), customTypeParams.toDriver, customTypeParams.fromDriver);
        }
    }(dbName);
}
exports.customType = customType;
