// drizzle-orm-pg/columns/interval.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.interval = exports.PgInterval = exports.PgIntervalBuilder = void 0;
const common_1 = require("./common");
class PgIntervalBuilder extends common_1.PgColumnBuilder {
    constructor(name, intervalConfig) {
        super(name);
        this.config.intervalConfig = intervalConfig;
    }
    /** @internal */
    build(table) {
        return new PgInterval(table, this.config);
    }
}
exports.PgIntervalBuilder = PgIntervalBuilder;
class PgInterval extends common_1.PgColumn {
    constructor(table, config) {
        super(table, config);
        this.config = config.intervalConfig;
    }
    getSQLType() {
        const fields = this.config.fields ? ` ${this.config.fields}` : '';
        const precision = this.config.precision ? ` (${this.config.precision})` : '';
        return `interval${fields}${precision}`;
    }
}
exports.PgInterval = PgInterval;
function interval(name, config = {}) {
    return new PgIntervalBuilder(name, config);
}
exports.interval = interval;
