// drizzle-orm-pg/columns/json.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.json = exports.PgJson = exports.PgJsonBuilder = void 0;
const common_1 = require("./common");
class PgJsonBuilder extends common_1.PgColumnBuilder {
    constructor(name) {
        super(name);
    }
    /** @internal */
    build(table) {
        return new PgJson(table, this.config);
    }
}
exports.PgJsonBuilder = PgJsonBuilder;
class PgJson extends common_1.PgColumn {
    constructor(table, config) {
        super(table, config);
    }
    getSQLType() {
        return 'json';
    }
    mapToDriverValue(value) {
        return JSON.stringify(value);
    }
    mapFromDriverValue(value) {
        if (typeof value === 'string') {
            try {
                return JSON.parse(value);
            }
            catch (e) {
                return value;
            }
        }
        return value;
    }
}
exports.PgJson = PgJson;
function json(name) {
    return new PgJsonBuilder(name);
}
exports.json = json;
