// drizzle-orm-pg/columns/jsonb.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.jsonb = exports.PgJsonb = exports.PgJsonbBuilder = void 0;
const common_1 = require("./common");
class PgJsonbBuilder extends common_1.PgColumnBuilder {
    constructor(name) {
        super(name);
    }
    /** @internal */
    build(table) {
        return new PgJsonb(table, this.config);
    }
}
exports.PgJsonbBuilder = PgJsonbBuilder;
class PgJsonb extends common_1.PgColumn {
    constructor(table, config) {
        super(table, config);
    }
    getSQLType() {
        return 'jsonb';
    }
    mapToDriverValue(value) {
        return JSON.stringify(value);
    }
    mapFromDriverValue(value) {
        if (typeof value === 'string') {
            try {
                return JSON.parse(value);
            }
            catch (e) {
                return value;
            }
        }
        return value;
    }
}
exports.PgJsonb = PgJsonb;
function jsonb(name) {
    return new PgJsonbBuilder(name);
}
exports.jsonb = jsonb;
