// drizzle-orm-pg/columns/numeric.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.decimal = exports.numeric = exports.PgNumeric = exports.PgNumericBuilder = void 0;
const common_1 = require("./common");
class PgNumericBuilder extends common_1.PgColumnBuilder {
    constructor(name, precision, scale) {
        super(name);
        this.config.precision = precision;
        this.config.scale = scale;
    }
    /** @internal */
    build(table) {
        return new PgNumeric(table, this.config);
    }
}
exports.PgNumericBuilder = PgNumericBuilder;
class PgNumeric extends common_1.PgColumn {
    constructor(table, config) {
        super(table, config);
        this.precision = config.precision;
        this.scale = config.scale;
    }
    getSQLType() {
        if (typeof this.precision !== 'undefined' && typeof this.scale !== 'undefined') {
            return `numeric(${this.precision}, ${this.scale})`;
        }
        else if (typeof this.precision === 'undefined') {
            return 'numeric';
        }
        else {
            return `numeric(${this.precision})`;
        }
    }
}
exports.PgNumeric = PgNumeric;
function numeric(name, config) {
    return new PgNumericBuilder(name, config === null || config === void 0 ? void 0 : config.precision, config === null || config === void 0 ? void 0 : config.scale);
}
exports.numeric = numeric;
exports.decimal = numeric;
