// drizzle-orm-pg/columns/time.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.time = exports.PgTime = exports.PgTimeBuilder = void 0;
const common_1 = require("./common");
const date_common_1 = require("./date.common");
class PgTimeBuilder extends date_common_1.PgDateColumnBaseBuilder {
    constructor(name, withTimezone, precision) {
        super(name);
        this.withTimezone = withTimezone;
        this.precision = precision;
    }
    /** @internal */
    build(table) {
        return new PgTime(table, this.config);
    }
}
exports.PgTimeBuilder = PgTimeBuilder;
class PgTime extends common_1.PgColumn {
    constructor(table, config) {
        super(table, config);
        this.withTimezone = config.withTimezone;
        this.precision = config.precision;
    }
    getSQLType() {
        const precision = typeof this.precision !== 'undefined' ? ` (${this.precision})` : '';
        return `time${precision}${this.withTimezone ? ' with time zone' : ''}`;
    }
}
exports.PgTime = PgTime;
function time(name, config = {}) {
    var _a;
    return new PgTimeBuilder(name, (_a = config.withTimezone) !== null && _a !== void 0 ? _a : false, config.precision);
}
exports.time = time;
