// drizzle-orm-pg/node/driver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.drizzle = exports.NodePgDriver = void 0;
const pg_1 = require("pg");
const db_1 = require("../db");
const dialect_1 = require("../dialect");
const session_1 = require("./session");
class NodePgDriver {
    constructor(client, dialect, options = {}) {
        this.client = client;
        this.dialect = dialect;
        this.options = options;
        this.initMappers();
    }
    createSession() {
        return new session_1.NodePgSession(this.client, this.dialect, { logger: this.options.logger });
    }
    initMappers() {
        pg_1.types.setTypeParser(pg_1.types.builtins.TIMESTAMPTZ, (val) => val);
        pg_1.types.setTypeParser(pg_1.types.builtins.TIMESTAMP, (val) => val);
        pg_1.types.setTypeParser(pg_1.types.builtins.DATE, (val) => val);
    }
}
exports.NodePgDriver = NodePgDriver;
function drizzle(client, config = {}) {
    const dialect = new dialect_1.PgDialect();
    const driver = new NodePgDriver(client, dialect, { logger: config.logger });
    const session = driver.createSession();
    return new db_1.PgDatabase(dialect, session);
}
exports.drizzle = drizzle;
