// drizzle-orm-mysql/db.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MySqlDatabase = void 0;
const query_builders_1 = require("./query-builders");
const table_1 = require("./table");
const utils_1 = require("./utils");
class MySqlDatabase {
    constructor(
    /** @internal */
    dialect, 
    /** @internal */
    session) {
        this.dialect = dialect;
        this.session = session;
    }
    select(from) {
        const fields = (0, utils_1.orderSelectedFields)(from[table_1.MySqlTable.Symbol.Columns]);
        return new query_builders_1.MySqlSelect(from, fields, this.session, this.dialect);
    }
    update(table) {
        return new query_builders_1.MySqlUpdateBuilder(table, this.session, this.dialect);
    }
    insert(table) {
        return new query_builders_1.MySqlInsertBuilder(table, this.session, this.dialect);
    }
    delete(table) {
        return new query_builders_1.MySqlDelete(table, this.session, this.dialect);
    }
    execute(query) {
        return this.session.execute(query.getSQL());
    }
}
exports.MySqlDatabase = MySqlDatabase;
