// drizzle-orm-mysql/indexes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.uniqueIndex = exports.index = exports.Index = exports.IndexBuilder = exports.IndexBuilderOn = void 0;
class IndexBuilderOn {
    constructor(name, unique) {
        this.name = name;
        this.unique = unique;
    }
    on(...columns) {
        return new IndexBuilder(this.name, columns, this.unique);
    }
}
exports.IndexBuilderOn = IndexBuilderOn;
class IndexBuilder {
    constructor(name, columns, unique) {
        this.config = {
            name,
            columns,
            unique,
        };
    }
    using(using) {
        this.config.using = using;
        return this;
    }
    algorythm(algorythm) {
        this.config.algorythm = algorythm;
        return this;
    }
    lock(lock) {
        this.config.lock = lock;
        return this;
    }
    /** @internal */
    build(table) {
        return new Index(this.config, table);
    }
}
exports.IndexBuilder = IndexBuilder;
class Index {
    constructor(config, table) {
        this.config = Object.assign(Object.assign({}, config), { table });
    }
}
exports.Index = Index;
function index(name) {
    return new IndexBuilderOn(name, false);
}
exports.index = index;
function uniqueIndex(name) {
    return new IndexBuilderOn(name, true);
}
exports.uniqueIndex = uniqueIndex;
