// drizzle-orm-mysql/columns/decimal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.decimal = exports.MySqlDecimal = exports.MySqlDecimalBuilder = void 0;
const common_1 = require("./common");
class MySqlDecimalBuilder extends common_1.MySqlColumnBuilderWithAutoIncrement {
    constructor(name, precision, scale) {
        super(name);
        this.config.precision = precision;
        this.config.scale = scale;
    }
    /** @internal */
    build(table) {
        return new MySqlDecimal(table, this.config);
    }
}
exports.MySqlDecimalBuilder = MySqlDecimalBuilder;
class MySqlDecimal extends common_1.MySqlColumnWithAutoIncrement {
    constructor(table, config) {
        super(table, config);
        this.precision = config.precision;
        this.scale = config.scale;
    }
    getSQLType() {
        if (typeof this.precision !== 'undefined' && typeof this.scale !== 'undefined') {
            return `decimal(${this.precision}, ${this.scale})`;
        }
        else if (typeof this.precision === 'undefined') {
            return 'decimal';
        }
        else {
            return `decimal(${this.precision})`;
        }
    }
}
exports.MySqlDecimal = MySqlDecimal;
function decimal(name, config = {}) {
    return new MySqlDecimalBuilder(name, config.precision, config.scale);
}
exports.decimal = decimal;
