// drizzle-orm-mysql/columns/float.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.float = exports.MySqlFloat = exports.MySqlFloatBuilder = void 0;
const common_1 = require("./common");
class MySqlFloatBuilder extends common_1.MySqlColumnBuilderWithAutoIncrement {
    constructor(name, precision, scale) {
        super(name);
        this.config.precision = precision;
        this.config.scale = scale;
    }
    /** @internal */
    build(table) {
        return new MySqlFloat(table, this.config);
    }
}
exports.MySqlFloatBuilder = MySqlFloatBuilder;
class MySqlFloat extends common_1.MySqlColumnWithAutoIncrement {
    constructor(table, config) {
        super(table, config);
        this.precision = config.precision;
        this.scale = config.scale;
    }
    getSQLType() {
        return 'float';
        // if (typeof this.precision !== 'undefined' && typeof this.scale !== 'undefined') {
        // 	return `float(${this.precision}, ${this.scale})`;
        // } else if (typeof this.precision === 'undefined') {
        // 	return 'float';
        // } else {
        // 	return `float(${this.precision})`;
        // }
    }
}
exports.MySqlFloat = MySqlFloat;
function float(name) {
    return new MySqlFloatBuilder(name);
}
exports.float = float;
