// drizzle-orm-mysql/columns/real.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.real = exports.MySqlReal = exports.MySqlRealBuilder = void 0;
const common_1 = require("./common");
class MySqlRealBuilder extends common_1.MySqlColumnBuilderWithAutoIncrement {
    constructor(name, precision, scale) {
        super(name);
        this.config.precision = precision;
        this.config.scale = scale;
    }
    /** @internal */
    build(table) {
        return new MySqlReal(table, this.config);
    }
}
exports.MySqlRealBuilder = MySqlRealBuilder;
class MySqlReal extends common_1.MySqlColumnWithAutoIncrement {
    constructor(table, config) {
        super(table, config);
        this.precision = config.precision;
        this.scale = config.scale;
    }
    getSQLType() {
        if (typeof this.precision !== 'undefined' && typeof this.scale !== 'undefined') {
            return `real(${this.precision}, ${this.scale})`;
        }
        else if (typeof this.precision === 'undefined') {
            return 'real';
        }
        else {
            return `real(${this.precision})`;
        }
    }
}
exports.MySqlReal = MySqlReal;
function real(name, config = {}) {
    return new MySqlRealBuilder(name, config.precision, config.scale);
}
exports.real = real;
