// drizzle-orm-mysql/columns/text.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.longtext = exports.mediumtext = exports.tinytext = exports.text = exports.MySqlText = exports.MySqlTextBuilder = void 0;
const common_1 = require("./common");
class MySqlTextBuilder extends common_1.MySqlColumnBuilder {
    constructor(name, textType) {
        super(name);
        this.config.textType = textType;
    }
    /** @internal */
    build(table) {
        return new MySqlText(table, this.config);
    }
}
exports.MySqlTextBuilder = MySqlTextBuilder;
class MySqlText extends common_1.MySqlColumn {
    constructor(table, config) {
        super(table, config);
        this.textType = config.textType;
    }
    getSQLType() {
        return this.textType;
    }
}
exports.MySqlText = MySqlText;
function text(name) {
    return new MySqlTextBuilder(name, 'text');
}
exports.text = text;
function tinytext(name) {
    return new MySqlTextBuilder(name, 'tinytext');
}
exports.tinytext = tinytext;
function mediumtext(name) {
    return new MySqlTextBuilder(name, 'mediumtext');
}
exports.mediumtext = mediumtext;
function longtext(name) {
    return new MySqlTextBuilder(name, 'longtext');
}
exports.longtext = longtext;
