// drizzle-orm-mysql/columns/timestamp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.timestamp = exports.MySqlTimestampString = exports.MySqlTimestampStringBuilder = exports.MySqlTimestamp = exports.MySqlTimestampBuilder = void 0;
const common_1 = require("./common");
const date_common_1 = require("./date.common");
class MySqlTimestampBuilder extends date_common_1.MySqlDateColumnBaseBuilder {
    constructor(name, fsp) {
        super(name);
        this.config.fsp = fsp;
    }
    /** @internal */
    build(table) {
        return new MySqlTimestamp(table, this.config);
    }
}
exports.MySqlTimestampBuilder = MySqlTimestampBuilder;
class MySqlTimestamp extends common_1.MySqlColumn {
    constructor(table, config) {
        super(table, config);
        this.fsp = config.fsp;
    }
    getSQLType() {
        const precision = typeof this.fsp !== 'undefined' ? ` (${this.fsp})` : '';
        return `timestamp${precision}`;
    }
    mapFromDriverValue(value) {
        return new Date(value);
    }
}
exports.MySqlTimestamp = MySqlTimestamp;
class MySqlTimestampStringBuilder extends date_common_1.MySqlDateColumnBaseBuilder {
    constructor(name, fsp) {
        super(name);
        this.config.fsp = fsp;
    }
    /** @internal */
    build(table) {
        return new MySqlTimestampString(table, this.config);
    }
}
exports.MySqlTimestampStringBuilder = MySqlTimestampStringBuilder;
class MySqlTimestampString extends common_1.MySqlColumn {
    constructor(table, config) {
        super(table, config);
        this.fsp = config.fsp;
    }
    getSQLType() {
        const precision = typeof this.fsp !== 'undefined' ? ` (${this.fsp})` : '';
        return `timestamp${precision}`;
    }
}
exports.MySqlTimestampString = MySqlTimestampString;
function timestamp(name, config) {
    if ((config === null || config === void 0 ? void 0 : config.mode) === 'string') {
        return new MySqlTimestampStringBuilder(name, config.fsp);
    }
    return new MySqlTimestampBuilder(name, config === null || config === void 0 ? void 0 : config.fsp);
}
exports.timestamp = timestamp;
